src/parsec/agent/runners/py/impl.cpp

```diff
@@ -0,0 +1,413 @@
+        return true;
+    }
+
+    void py_runner::parse_header() {
+        /* Assumes that header is <n returns> <n inputs> | return types |
+           return args | input types | input args | function code */
+        auto functionString
+            = std::string(static_cast<char*>(m_function.data()));
+
+        m_input_args.clear();
+        m_return_args.clear();
+        m_return_types.clear();
+
+        auto arg_delim = functionString.find('|');
+        auto arg_string = functionString.substr(0, arg_delim);
+        size_t pos = 0;
+        m_return_types = arg_string;
+        functionString.erase(0, arg_delim + 1);
+
+        arg_delim = functionString.find('|');
+        arg_string = functionString.substr(0, arg_delim);
+        while((pos = arg_string.find(',', 0)) != std::string::npos) {
+            m_return_args.push_back(arg_string.substr(0, pos));
+            arg_string.erase(0, pos + 1);
+        }
+        functionString.erase(0, arg_delim + 1);
+
+        arg_delim = functionString.find('|');
+        arg_string = functionString.substr(0, arg_delim);
+        while((pos = arg_string.find(',', 0)) != std::string::npos) {
+            m_input_args.push_back(arg_string.substr(0, pos));
+            arg_string.erase(0, pos + 1);
+        }
+        functionString.erase(0, arg_delim + 1);
+
+        m_function = cbdc::buffer();
+        m_function.append(functionString.c_str(), functionString.size());
+        m_function.append("\0", 1);
+    }
+
+    auto py_runner::parse_params() -> std::vector<std::string> {
+        std::vector<std::string> params(0);
+
+        // Input validation to ensure that it's safe to use strlen
+        if(m_param.size() == 0) {
+            m_log->error("m_param contains no data");
+            return params;
+        }
+        m_param.append("\0", 1);
+        size_t pipeCount = 0;
+        size_t stringCount = 0;
+        auto paramString
+            = std::string(static_cast<char*>(m_param.data()), m_param.size());
+        for(size_t i = 0; i < m_param.size(); i++) {
+            if(paramString[i] == '|') {
+                pipeCount++;
+            } else if(i > 0 && paramString[i] == '\0'
+                      && paramString[i - 1] != '\0') {
+                stringCount++;
+            }
+        }
+        auto bail = false;
+        if(pipeCount != 3) {
+            m_log->error("m_param sections are improperly formatted");
+            bail = true;
+        } else if(stringCount
+                  != m_input_args.size() + m_return_args.size() + pipeCount) {
+            m_log->error("m_param contains too few arguments or arguments are "
+                         "improperly formatted");
+            bail = true;
+        }
+        if(bail) {
+            return params;
+        }
+        // ---
+
+        // get parameters that are user inputs
+        size_t bufferOffset = 0;
+        auto it = 0;
+        // Known that there are 3 parts of the header
```

Comment:
This feels a touch magical. Is this guaranteed, python-version-dependent, or internally-specified?

If internally-specified, is there benefit to pulling these cases out as an enum?

Reply:
This is internally specified, and the input validation is handled just above this code block -- just to check, are you referring to the cases in the switch statement below this block?

Reply:
Yeah, the actual line I highlighted for this comment was line 194 (the comment asserting that there are three parts of the header). 👍